projects/utils-maps/src/lib/ol/utils-ol-mapstate.ts

```diff
@@ -0,0 +1,204 @@
+  } else {
+    return extent;
+  }
+}
+
+/**
+ * @param geographic: boolean
+ * @returns olExtend: [minX, minY, maxX, maxY]
+ */
+export function getCurrentExtent(map: Map, geographic?: boolean) {
+  const projection = (geographic) ? getProjection('EPSG:4326') : getProjection(DEFAULT_PROJECTION);
+  const extent = map.getView().calculateExtent();
+  const transformedExtent = transformExtent(extent, map.getView().getProjection().getCode(), projection);
+  return (transformedExtent as Extent);
+}
+
+
+export function setZoom(map: Map, zoom: number, notifier?: 'map' | 'user') {
+  const view = map.getView();
+  view.setZoom(zoom);
+}
+
+/** USED in map-ol.component */
+export function getZoom(map: Map): number {
+  return map.getView().getZoom();
+}
+
+export function zoomInOut(map: Map, value: '-' | '+') {
+  const view = map.getView();
+  if (!view) {
+    // the map does not have a view, so we can't act
+    // upon it
+    return;
+  }
+  const delta = 1, duration = 250;
+  const currentZoom = view.getZoom();
+  if (currentZoom !== undefined) {
+    const newZoom = view.getConstrainedZoom(currentZoom + delta);
+    if (duration > 0) {
+      if (view.getAnimating()) {
+        view.cancelAnimations();
+      }
+      view.animate({
+        zoom: newZoom,
+        duration,
+        easing: easeOut
+      });
+    } else {
+      view.setZoom(newZoom);
+    }
+  }
+}
+
+
+
+export function getMapProjection(map: Map) {
+  return map.getView().getProjection();
+}
+
+
+/**
+ * vector layers will be reprojected automatically
+ * wms layers will be updated with corresponding proj def in the requests.
+ * for other raster layers and for those wms layers whose backend does not support target projection, please
+ * define initial(default) layer projection, so openlayers will reproject on the client side
+ * projection is proj~ProjectionLike
+ */
+export function setMapProjection(map: Map, projection: Projection | string, viewOptions?: ViewOptions, cb?: (projection: Projection) => void) {
+  let epsg = DEFAULT_PROJECTION;
+  let view = null;
+  if (projection) {
+    let newViewOptions: ViewOptions = {};
+    if (viewOptions) {
+      newViewOptions = viewOptions;
+      newViewOptions.minResolution = undefined;
+      newViewOptions.maxResolution = undefined;
+      newViewOptions.resolution = undefined;
+      newViewOptions.resolutions = undefined;
+    }
+    if (projection instanceof Projection) {
+      newViewOptions.projection = projection;
+      const newCenter = transform(map.getView().getCenter(), map.getView().getProjection(), projection); // get center coordinates in the new projection
+      newViewOptions.center = newCenter; // map.getView().getCenter();
+      // _viewOptions.extent = projection.getExtent();// || undefined;
+      newViewOptions.zoom = map.getView().getZoom();
+    } else if (typeof projection === 'string') {
+      newViewOptions.projection = projection;
+      newViewOptions.center = map.getView().getCenter();
```

Comment:
Here, the new center is not reprojected - contrary to what we did in line 145. I'm quite certain that this is a mistake. I'll change the code accordingly, would that be ok?

Reply:
yes, but be carefully, in this situation the projection is defined as a string code (e.g. "EPSG:4326"), I am not sure if transform will accept this.